### PR TITLE
Ruby 2.3

### PR DESCRIPTION
**Pull request by KnVerey**
Upgrades our minimum requirement to Ruby 2.3. Also makes a few easy-win changes to 2.3-style squiggly heredocs, `dig` and `&.`.

The new `# frozen_string_literal: true` came from running `bx rubocop -a`. Dunno why it didn't care about those files until now.

Closes https://github.com/Shopify/kubernetes-deploy/issues/134

### CODE REVIEW

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb [hunk 1]
```diff
@@ -86,7 +86,7 @@ def unmanaged?
 
     def container_names
       regular_containers = @definition["spec"]["template"]["spec"]["containers"].map { |c| c["name"] }
-      init_containers = @definition["spec"]["template"]["spec"].fetch("initContainers", {}).map { |c| c["name"] }
+      init_containers = @definition["spec"]["template"]["spec"].fetch("initContainers", []).map { |c| c["name"] }
```
Comment (stefanmb) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: Will this still work if the default value is an array instead of a dictionary?

File: .rubocop.yml [hunk 2]
```diff
@@ -2,7 +2,4 @@ inherit_from:
   - http://shopify.github.io/ruby-style-guide/rubocop.yml
 
 AllCops:
-  TargetRubyVersion: 2.1
-
-Layout/IndentHeredoc:
-  EnforcedStyle: active_support
+  TargetRubyVersion: 2.3
```
Comment (kirs) [inline .rubocop.yml]: > The new # frozen_string_literal: true came from running bx rubocop -a. Dunno why it didn't care about those files until now.

Because it has been validating code for 2.1, and in that version there was no such thing as frozen string literal

File: kubernetes-deploy.gemspec [hunk 3]
```diff
@@ -20,6 +21,8 @@ Gem::Specification.new do |spec|
   spec.bindir        = "exe"
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ["lib"]
+
+  spec.required_ruby_version = '>= 2.3.0'
```
Comment (kirs) [inline kubernetes-deploy.gemspec]: 👍

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 4]
```diff
@@ -47,15 +47,15 @@ def deploy_failed?
     end
 
     def failure_message
-      @latest_rs.failure_message
+      @latest_rs&.failure_message
```
Comment (kirs) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Why this change if it worked before?

File: test/unit/kubernetes-deploy/google_friendly_config_test.rb [hunk 5]
```diff
@@ -33,7 +33,7 @@ def test_auth_use_default_gcp_success
   def test_auth_use_default_gcp_failure
     config = KubernetesDeploy::KubeclientBuilder::GoogleFriendlyConfig.new(kubeconfig, "")
 
-    stub_request(:post, 'https://www.googleapis.com/oauth2/v3/token')
+    stub_request(:post, 'https://www.googleapis.com/oauth2/v4/token')
```
Comment (kirs) [inline test/unit/kubernetes-deploy/google_friendly_config_test.rb]: Curious why this is necessary if dependencies weren't updated.

File: .rubocop.yml (hunk 2 above)
Comment (KnVerey) [inline .rubocop.yml]: Ah, of course! I could have sworn it had bugged me to add some of those in the past, but I must be misremembering. Or maybe I ran it with the wrong config somehow. ¯\\\_(ツ)_/¯

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb (hunk 4 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: It is unlikely but possible that no replicaset will be found. That's why we have `return false unless @latest_rs` in other spots. I'm not sure what made me notice these guys now, but we shouldn't assume we have an RS here either.

File: lib/kubernetes-deploy/kubernetes_resource/replica_set.rb (hunk 1 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource/replica_set.rb]: yeah, `{}.map` and `[].map` both have the same result--do nothing and return an array.

image

I changed it because `initContainers` would be an array if present

File: test/unit/kubernetes-deploy/google_friendly_config_test.rb (hunk 5 above)
Comment (KnVerey) [inline test/unit/kubernetes-deploy/google_friendly_config_test.rb]: Stefan [asked the same thing in the other PR](https://github.com/Shopify/kubernetes-deploy/pull/141#discussion_r128847918), which has a copy of this commit (CI won't pass without it or an equivalent fix):

> Since this is a gem, the Gemfile.lock isn't committed to the repo. Our CI bundle installs from scratch every time, so it'll upgrade as far as the Gemfile allows it to when new versions become available. Because of this, CI just started using v0.4.0 of kubeclient, which is what caused this difference. Alternatively, we could pin kubeclient more strictly in the gemspec.

LMK if you think we should change the gemspec instead.
